File: InversionCount/InversionCount.cpp
// InversionCount.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
#include <vector>

int GetInversionCount(const std::vector<int>& arr)
{
    int count = 0;
    for (auto i = arr.cbegin(); i != arr.cend(); ++i)
    {
        for (auto j = i + 1; j != arr.cend(); ++j)
        {
            if (*i > *j)
                ++count;
        }
    }

    return count;
}

int main()
{
    while (true)
    {
        int count = 0;
        std::cout << "Number of elements in the array: ";
        std::cin >> count;
        if (count == 0)
            break;

        std::vector<int> vec(count);
        std::cout << "The number array: ";
        for (int i = 0; i < count; ++i)
            std::cin >> vec[i];

        std::cout << "The inversion count of the array: " << GetInversionCount(vec) << std::endl << std::endl;
    }
}

File: Go/InversionCount.go
package main

import "fmt"

func InversionCount(arr []int, index int) int {
	count := 0
	for i := index + 1; i < len(arr); i++ {
		if arr[i] < arr[index] {
			count++
		}
	}

	return count
}

func GetInversionCount(arr []int) int {
	count := 0
	for i := 0; i < len(arr)-1; i++ {
		count += InversionCount(arr, i)
	}

	return count
}

func main() {
	for {
		fmt.Print("Number of elements: ")
		count := 0
		fmt.Scan(&count)
		if count == 0 {
			break
		}

		fmt.Print("Enter the array: ")
		vec := make([]int, count)
		for i := 0; i < count; i++ {
			fmt.Scan(&vec[i])
		}

		fmt.Printf("To sort the array, it needs %d moves\n", GetInversionCount(vec))
	}
}
